packages/itwin/grouping-mapping-widget/src/components/Mappings/Extraction/ExtractionMessageModal.tsx

```diff
@@ -24,58 +32,75 @@ export interface ExtractionMessageModalProps {
 }
 
 export const ExtractionMessageModal = ({ isOpen, onClose, extractionMessageData, timestamp }: ExtractionMessageModalProps) => {
-  const [formattedExtractionMessage, setFormattedExtractionMessage] = useState<ExtractionMessageData[]>([]);
+  const [formattedExtractionMessage, setFormattedExtractionMessage] = useState<ExtractionMessageData[] | undefined>(undefined);
   const groupingMappingApiConfig = useGroupingMappingApiConfig();
   const mappingClient = useMappingClient();
-  const { mappings } = useMappingsOperations({...groupingMappingApiConfig, mappingClient});
+  const { mappings, isLoading: isMappingsLoading } = useMappingsOperations({ ...groupingMappingApiConfig, mappingClient });
   const [formattedTimestamp, setFormattedTimestamp] = useState<string>("");
 
-  const getMappingName = useCallback(async (mappingId: string) => {
-    return mappings.find((mapping) => {return mapping.id === mappingId;})?.mappingName ?? "";
-  }, [mappings]);
-
-  const getGroupNames = useCallback(async (mappingId: string, groupId: string, groupsCache: Map<string, Group[]>) => {
-    const accessToken = await groupingMappingApiConfig.getAccessToken();
-    if(!groupsCache.has(mappingId)){
-      const groups = await mappingClient.getGroups(
-        accessToken,
-        groupingMappingApiConfig.iModelId,
-        mappingId
-      );
-      groupsCache.set(mappingId, groups);
-    }
-    return groupsCache.get(mappingId)?.find((group) => {return group.id === groupId;})?.groupName ?? "";
-  }, [groupingMappingApiConfig, mappingClient]);
+  const getMappingName = useCallback((mappingId: string, mappings: Mapping[]) => {
+    return mappings.find((mapping) => mapping.id === mappingId)?.mappingName ?? "";
+  }, []);
+
+  // Extract groupIds and mappingIds from messages
+  const extractionInfo = useMemo(() => {
+    return extractionMessageData.reduce<{
+      mappingId: string;
+      groupId: string;
+    }[]>((acc, message) => {
+      const splittedMessage = message.message.split(" ");
+      const mappingId = splittedMessage[splittedMessage.indexOf("MappingId:") + 1]?.match(/^([^,]+),$/)?.[1];
+      const groupId = splittedMessage[splittedMessage.indexOf("GroupId:") + 1]?.match(/^([^,]+).$/)?.[1];
+
+      if (mappingId && groupId) {
+        acc.push({ mappingId, groupId });
+      }
+      return acc;
+    }, []);
+  }, [extractionMessageData]);
+
+  // useQueries to fetch all group names
+  const groupQueriesResults = useQueries({
+    queries: extractionInfo.map(({ mappingId, groupId }) => ({
+      queryKey: ["groups", mappingId, groupId],
+      queryFn: async () => {
+        const accessToken = await groupingMappingApiConfig.getAccessToken();
+        const groups = await mappingClient.getGroups(
```

Comment:
How many requests does this result in if there is more than 1 unique group in the extractionInfo array?

Reply:
I just rewrote it to just fetching groups. Added in a new hook from the work I've been doing with the groups components.